src/wallet.rs

```diff
@@ -194,6 +205,185 @@ impl Wallet {
+        self.force = overwrite;
+        self
+    }
+
+    /// Set the seed type if using seed words to create this wallet
+    /// Defaults to None
+    pub fn seed_type(mut self, seed_type: Option<SeedType>) -> Builder {
+        self.seed_type = seed_type;
+        self
+    }
+
+    /// The seed words used to create this wallet
+    /// Defaults to None
+    pub fn seed_words(mut self, seed_words: Option<Vec<String>>) -> Builder {
+        self.seed_words = seed_words;
+        self
+    }
+
+    /// The type of key to generate (ecc_compact/ed25519)
+    /// Defaults to ed25519
+    pub fn key_tag(mut self, key_tag: &KeyTag) -> Builder {
+        self.key_tag = *key_tag;
+        self
+    }
+
+    /// Optional shard config info to use in order to create a sharded wallet
+    /// otherwise, creates a basic non-sharded wallet
+    pub fn shard(mut self, shard_config: Option<ShardConfig>) -> Builder {
+        self.shard = shard_config;
+        self
+    }
+
+    /// Creates a new wallet
+    pub fn create(self) -> Result<Wallet> {
+        let keypair = gen_keypair(self.key_tag, self.seed_words, self.seed_type.as_ref())?;
+
+        let wallet = if let Some(shard_config) = &self.shard {
+            let format = format::Sharded {
+                key_share_count: shard_config.key_share_count,
+                recovery_threshold: shard_config.recovery_threshold,
+                pwhash: self.pwhash,
+                key_shares: vec![],
+            };
+            Wallet::encrypt(&keypair, self.password.as_bytes(), Format::Sharded(format))?
+        } else {
+            let format = format::Basic {
+                pwhash: PwHash::argon2id13_default(),
+            };
+            Wallet::encrypt(&keypair, self.password.as_bytes(), Format::Basic(format))?
+        };
+
+        if self.shard.is_some() {
+            let extension = self
+                .output
+                .extension()
+                .unwrap_or_else(|| OsStr::new(""))
+                .to_str()
+                .unwrap()
+                .to_string();
+            for (i, shard) in wallet.shards()?.iter().enumerate() {
+                let mut filename = self.output.clone();
+                let share_extension = format!("{}.{}", extension, (i + 1).to_string());
+                filename.set_extension(share_extension);
+                let mut writer = open_output_file(&filename, !self.force)?;
+                shard.write(&mut writer)?;
+            }
+        } else {
+            let mut writer = open_output_file(&self.output, !self.force)?;
+            wallet.write(&mut writer)?;
+        }
+
+        Ok(wallet)
+    }
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+fn gen_keypair(
+    tag: KeyTag,
+    seed_words: Option<Vec<String>>,
+    seed_type: Option<&SeedType>,
+) -> Result<Keypair> {
+    // Callers of this function should either have Some of both or None of both.
```

Comment:
hmm why not take an `Option<(Vec<String>, &SeedType)>` if they both have to be there?